File: docs/validation/go/app/scenarios/av08_version_fidelity.go
package scenarios

// AV-08: File_Version string survives write->read unchanged

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	mc "machine-config-go"
)

func RunAv08VersionFidelity(fixturesDir, _ string) (bool, string) {
	path := filepath.Join(fixturesDir, "reference_config.h5")
	cfg, err := mc.NewReader(path).Parse()
	if err != nil {
		return false, fmt.Sprintf("read failed: %v", err)
	}
	origVersion := strings.TrimSpace(cfg.Meta.FileVersion)

	tmpPath, err := tempH5("mcl_go_av08")
	if err != nil {
		return false, fmt.Sprintf("tempfile: %v", err)
	}
	defer os.Remove(tmpPath)

	if err := mc.NewWriter().Write(cfg, tmpPath); err != nil {
		return false, fmt.Sprintf("write failed: %v", err)
	}
	rb, err := mc.NewReader(tmpPath).Parse()
	if err != nil {
		return false, fmt.Sprintf("readback failed: %v", err)
	}
	rbVersion := strings.TrimSpace(rb.Meta.FileVersion)

	if rbVersion != "1.0" {
		return false, fmt.Sprintf("file_version after roundtrip: expected '1.0', got %q", rbVersion)
	}
	if rbVersion != origVersion {
		return false, fmt.Sprintf("file_version changed: %q -> %q", origVersion, rbVersion)
	}

	return true, fmt.Sprintf("File_Version survives roundtrip unchanged: %q", rbVersion)
}

File: docs/validation/rust/app/src/scenarios/av08_version_fidelity.rs
// AV-08: File_Version string survives write→read unchanged
use machine_config::{MachineConfigReader, MachineConfigWriter};
use std::path::Path;

pub fn run(fixtures_dir: &Path, _real_dir: &Path) -> (bool, String) {
    let path = fixtures_dir.join("reference_config.h5");
    let cfg = match MachineConfigReader::open(&path).and_then(|r| r.parse()) {
        Ok(c) => c,
        Err(e) => return (false, format!("read failed: {e}")),
    };
    let orig_version = cfg.meta.file_version.trim().to_string();

    let tmp = match tempfile::Builder::new().suffix(".h5").tempfile() {
        Ok(t) => t,
        Err(e) => return (false, format!("tempfile creation failed: {e}")),
    };
    if let Err(e) = MachineConfigWriter::new(&cfg).write(tmp.path()) {
        return (false, format!("write failed: {e}"));
    }
    let rb = match MachineConfigReader::open(tmp.path()).and_then(|r| r.parse()) {
        Ok(c) => c,
        Err(e) => return (false, format!("readback failed: {e}")),
    };
    let rb_version = rb.meta.file_version.trim().to_string();

    if rb_version != "1.0" {
        return (false, format!("file_version after roundtrip: expected '1.0', got '{rb_version}'"));
    }
    if rb_version != orig_version {
        return (false, format!("file_version changed: '{orig_version}' -> '{rb_version}'"));
    }

    (true, format!("File_Version survives roundtrip unchanged: '{rb_version}'"))
}
